cmd/containerboot/ingressservices.go

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build linux
+
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/netip"
+	"os"
+	"path/filepath"
+	"reflect"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"tailscale.com/kube/ingressservices"
+	"tailscale.com/kube/kubeclient"
+	"tailscale.com/util/linuxfw"
+	"tailscale.com/util/mak"
+)
+
+// ingressProxy corresponds to a Kubernetes Operator's network layer ingress
+// proxy. It configures firewall rules (iptables or nftables) to proxy tailnet
+// traffic to Kubernetes Services.  Currently this is only used for network
+// layer proxies in HA mode.
+type ingressProxy struct {
+	cfgPath string // path to ingress configfile.
+
+	// nfr is the netfilter runner used to configure firewall rules.
+	// This is going to be either iptables or nftables based runner.
+	// Never nil.
+	nfr linuxfw.NetfilterRunner
+
+	kc          kubeclient.Client // never nil
+	stateSecret string            // Secret that holds Tailscale state
+
+	// Pod's IP addresses are used as an identifier of this particular Pod.
+	podIPv4 string // empty if Pod does not have IPv4 address
+	podIPv6 string // empty if Pod does not have IPv6 address
+}
+
+// run starts the ingress proxy and ensures that firewall rules are set on start
+// and refreshed as ingress config changes.
+func (ep *ingressProxy) run(ctx context.Context, opts ingressProxyOpts) error {
```

Comment:
nit: but maybe `ip` is too overloaded so just `i` or `p`?

```suggestion
func (ip *ingressProxy) run(ctx context.Context, opts ingressProxyOpts) error {
```

Reply:
`p` sgtm